File: android/core/network/src/main/kotlin/xyz/klinik/network/PhotosApi.kt
package xyz.klinik.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
enum class PhotoCategory {
    BEFORE,
    AFTER,
    COMPLICATION,
    WOUND,
    ;

    val stringKey: String get() = "photo_category_${name.lowercase()}"
}

@Serializable
data class ClinicalPhoto(
    val id: String,
    val category: PhotoCategory,
    val bodyArea: String? = null,
    /** Free text: milestones differ per procedure (pre-op, post-op D1, W2, M1…). */
    val phaseLabel: String? = null,
    val mime: String,
    val size: Int,
    val takenAt: String,
    /** Location metadata was removed before the image was stored. */
    val exifStripped: Boolean = false,
    val isFaceBlurred: Boolean = false,
    /** Without a photo-usage consent the image is clinical-use only. */
    val consentId: String? = null,
    val note: String? = null,
    /**
     * AI pre-assessment (spec M5). Null means nobody has looked.
     *
     * A **flag**, never a diagnosis, and never shown to a patient — the photo
     * endpoints require `photos.read`, which no patient holds.
     */
    val aiReviewSuggested: Boolean? = null,
    /** What was observed, from a closed vocabulary. Never a condition name. */
    val aiFindings: List<String> = emptyList(),
    val aiAssessedAt: String? = null,
) {
    val hasUsageConsent: Boolean get() = consentId != null

    /** Somebody looked and found nothing, as opposed to nobody having looked. */
    val isAssessedClean: Boolean get() = aiReviewSuggested == false
    val needsReview: Boolean get() = aiReviewSuggested == true

    fun findingKeys(): List<String> = aiFindings.map { "photo_finding_${it.replace('-', '_')}" }
}

/** Why nothing was assessed, when nothing was. */
@Serializable
enum class AssessmentSkip {
    /** The clinic has not switched photo assessment on. */
    @kotlinx.serialization.SerialName("disabled")
    DISABLED,

    @kotlinx.serialization.SerialName("unsupported-image")
    UNSUPPORTED_IMAGE,

    @kotlinx.serialization.SerialName("ai-unavailable")
    AI_UNAVAILABLE,

    /** The answer could not be read, so the photo is left exactly as it was. */
    @kotlinx.serialization.SerialName("unreadable")
    UNREADABLE,
}

@Serializable
data class PhotoAssessment(
    val photo: ClinicalPhoto,
    /** From a closed vocabulary. A word outside it is dropped, not passed through. */
    val findings: List<String> = emptyList(),
    /** Any finding at all means a clinician should look. */
    val reviewSuggested: Boolean = false,
    val model: String? = null,
    val skippedReason: AssessmentSkip? = null,
) {
    val wasAssessed: Boolean get() = skippedReason == null
}

@Serializable
data class GalleryGroup(
    val bodyArea: String? = null,
    /** Oldest first: a progression reads forwards. */
    val photos: List<ClinicalPhoto> = emptyList(),
) {
    val id: String get() = bodyArea ?: ""
}

@Serializable
data class PhotoLink(val url: String, val expiresAt: String)

class PhotosApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    /**
     * Photos the pre-assessment flagged, oldest first.
     *
     * A worklist: ordered newest-first it would be one where the oldest thing
     * waits forever.
     */
    suspend fun flagged(): List<ClinicalPhoto> =
        decode(client.send(Endpoint(HttpMethod.GET, "photos/flagged")))

    /** Asks for a pre-assessment. Sends a clinical photograph to a third party. */
    suspend fun assess(photoId: String): PhotoAssessment =
        decode(client.send(Endpoint(HttpMethod.POST, "photos/$photoId/assess")))

    suspend fun gallery(
        subject: RecordSubject,
        category: PhotoCategory? = null,
        bodyArea: String? = null,
    ): List<GalleryGroup> =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.GET,
                    subject.base("photos"),
                    query = buildMap {
                        category?.let { put("category", it.name) }
                        bodyArea?.let { put("bodyArea", it) }
                    },
                ),
            ),
        )

    /**
     * The photo a new capture should be lined up against.
     *
     * The endpoint answers with an empty object when there is nothing to line
     * up against, which is not the same as a missing field or a failure.
     */
    suspend fun overlayReference(subject: RecordSubject, bodyArea: String): ClinicalPhoto? {
        val body = client.send(
            Endpoint(
                HttpMethod.GET,
                subject.base("photos/overlay"),
                query = mapOf("bodyArea" to bodyArea),
            ),
        )

        return runCatching { json.decodeFromString<ClinicalPhoto>(body) }.getOrNull()
    }

    /** A fresh link each time: they are short-lived and are never stored. */
    suspend fun link(photoId: String): PhotoLink =
        decode(client.send(Endpoint(HttpMethod.GET, "photos/$photoId/url")))

    suspend fun upload(
        subject: RecordSubject,
        path: String,
        filename: String,
        category: PhotoCategory,
        bodyArea: String? = null,
        phaseLabel: String? = null,
        note: String? = null,
        consentId: String? = null,
    ): ClinicalPhoto =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.POST,
                    subject.base("photos"),
                    multipart = MultipartUpload(
                        fields = buildMap {
                            put("category", category.name)
                            bodyArea?.let { put("bodyArea", it) }
                            phaseLabel?.let { put("phaseLabel", it) }
                            note?.let { put("note", it) }
                            consentId?.let { put("consentId", it) }
                        },
                        path = path,
                        filename = filename,
                        contentType = "image/jpeg",
                    ),
                ),
            ),
        )

    suspend fun remove(photoId: String) {
        client.send(Endpoint(HttpMethod.DELETE, "photos/$photoId"))
    }

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }
}

File: ios/Sources/KlinikAPI/PhotosAPI.swift
import Foundation
import KlinikCore

public enum PhotoCategory: String, Decodable, Encodable, Sendable, CaseIterable {
    case before = "BEFORE"
    case after = "AFTER"
    case complication = "COMPLICATION"
    case wound = "WOUND"

    public var localizedName: String { L10n.string("photo.category.\(rawValue)") }
}

public struct ClinicalPhoto: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    public let category: PhotoCategory
    public let bodyArea: String?
    /// Free text: milestones differ per procedure (pre-op, post-op D1, W2, M1…).
    public let phaseLabel: String?
    public let mime: String
    public let size: Int
    public let takenAt: Date
    /// Location metadata was removed before the image was stored.
    public let exifStripped: Bool
    public let isFaceBlurred: Bool
    /// Without a photo-usage consent the image is clinical-use only.
    public let consentId: String?
    public let note: String?

    /**
     * AI pre-assessment (spec M5). Null means nobody has looked.
     *
     * A **flag**, never a diagnosis, and never shown to a patient — the photo
     * endpoints require `photos.read`, which no patient holds.
     */
    public let aiReviewSuggested: Bool?
    /// What was observed, from a closed vocabulary. Never a condition name.
    public let aiFindings: [String]
    public let aiAssessedAt: Date?

    public var hasUsageConsent: Bool { consentId != nil }

    /// Somebody looked and found nothing, as opposed to nobody having looked.
    public var isAssessedClean: Bool { aiReviewSuggested == false }
    public var needsReview: Bool { aiReviewSuggested == true }

    public var localizedFindings: [String] {
        aiFindings.map { L10n.string("photo.finding.\($0)") }
    }
}

public struct GalleryGroup: Decodable, Sendable, Equatable, Identifiable {
    public let bodyArea: String?
    /// Oldest first: a progression reads forwards.
    public let photos: [ClinicalPhoto]

    public var id: String { bodyArea ?? "" }
}

/// Why nothing was assessed, when nothing was.
public enum AssessmentSkip: String, Decodable, Sendable, Equatable {
    /// The clinic has not switched photo assessment on.
    case disabled
    case unsupportedImage = "unsupported-image"
    case aiUnavailable = "ai-unavailable"
    /// The answer could not be read, so the photo is left exactly as it was.
    case unreadable
}

public struct PhotoAssessment: Decodable, Sendable, Equatable {
    public let photo: ClinicalPhoto
    /// From a closed vocabulary. A word outside it is dropped, not passed through.
    public let findings: [String]
    /// Any finding at all means a clinician should look.
    public let reviewSuggested: Bool
    public let model: String?
    public let skippedReason: AssessmentSkip?

    public var wasAssessed: Bool { skippedReason == nil }
}

public struct PhotoLink: Decodable, Sendable, Equatable {
    public let url: String
    public let expiresAt: Date
}

public struct PhotosAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    /**
     * Photos the pre-assessment flagged, oldest first.
     *
     * A worklist: ordered newest-first it would be one where the oldest thing
     * waits forever.
     */
    public func flagged() async throws -> [ClinicalPhoto] {
        try await client.send(
            Endpoint(method: .get, path: "photos/flagged"),
            as: [ClinicalPhoto].self
        )
    }

    /// Asks for a pre-assessment. Sends a clinical photograph to a third party.
    public func assess(_ photoId: String) async throws -> PhotoAssessment {
        try await client.send(
            Endpoint(method: .post, path: "photos/\(photoId)/assess"),
            as: PhotoAssessment.self
        )
    }

    public func gallery(
        subject: RecordSubject,
        category: PhotoCategory? = nil,
        bodyArea: String? = nil
    ) async throws -> [GalleryGroup] {
        var query: [String: String] = [:]
        if let category { query["category"] = category.rawValue }
        if let bodyArea { query["bodyArea"] = bodyArea }

        return try await client.send(
            Endpoint(method: .get, path: subject.base("photos"), query: query),
            as: [GalleryGroup].self
        )
    }

    /// The photo a new capture should be lined up against, or nil for the first
    /// of its body area.
    public func overlayReference(
        subject: RecordSubject,
        bodyArea: String
    ) async throws -> ClinicalPhoto? {
        let photo = try await client.send(
            Endpoint(
                method: .get,
                path: subject.base("photos/overlay"),
                query: ["bodyArea": bodyArea]
            ),
            as: OptionalPhoto.self
        )

        return photo.value
    }

    /// A fresh link each time: they are short-lived and are never stored.
    public func link(photoId: String) async throws -> PhotoLink {
        try await client.send(
            Endpoint(method: .get, path: "photos/\(photoId)/url"),
            as: PhotoLink.self
        )
    }

    public func upload(
        subject: RecordSubject,
        fileURL: URL,
        category: PhotoCategory,
        bodyArea: String?,
        phaseLabel: String?,
        takenAt: Date? = nil,
        note: String? = nil,
        consentId: String? = nil
    ) async throws -> ClinicalPhoto {
        var fields = ["category": category.rawValue]
        if let bodyArea { fields["bodyArea"] = bodyArea }
        if let phaseLabel { fields["phaseLabel"] = phaseLabel }
        if let takenAt { fields["takenAt"] = ISO8601DateFormatter().string(from: takenAt) }
        if let note { fields["note"] = note }
        if let consentId { fields["consentId"] = consentId }

        return try await client.upload(
            Endpoint(method: .post, path: subject.base("photos")),
            multipart: MultipartBody(
                fields: fields,
                fileURL: fileURL,
                contentType: "image/jpeg"
            ),
            as: ClinicalPhoto.self
        )
    }

    public func remove(photoId: String) async throws {
        try await client.send(Endpoint(method: .delete, path: "photos/\(photoId)"))
    }
}

/// The overlay endpoint answers with an empty object when there is nothing to
/// line up against, which is not the same as a missing field.
private struct OptionalPhoto: Decodable, Sendable {
    let value: ClinicalPhoto?

    init(from decoder: Decoder) throws {
        value = try? ClinicalPhoto(from: decoder)
    }
}
